sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java

```diff
@@ -0,0 +1,46 @@
+package com.mesosphere.sdk.dcos.auth;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Optional;
+
+/**
+ * CachedTokenProvider retrieves token from underlying provider and caches the value. It automatically triggers
+ * getToken() method on underlying provider when token is about to expire.
+ *
+ * // TODO(mh): This could be also running in a separate thread in background.
+ * // TODO(mh): Does this needs to be threadsafe?
```

Comment:
@benclarkwood you are right. My concern was that multiple objects would try to get access and possibly refresh authentication token that would lead to a race condition.